lib/event_queue.nit

```diff
@@ -0,0 +1,337 @@
+	private var actives = new Array[EventInfo[E]]
+
+	# List of previously active event
+	private var old_actives = new Array[EventInfo[E]]
+
+	# Global time since the creation of the queue
+	var time = 0.0
+
+	# Nearest deadline, used to optimise queue access
+	#
+	# `inf` if no deadline is set
+	private var next: Float = inf
+
+	# Register an `event` in the queue to be active in `delay` units of time.
+	#
+	# If `duration` is given, this is duration after what the event is automatically discarded.
+	# If `null`, 0.0 or not given, the event will be executed only once
+	# Use `inf` for an event with an infinite duration.
+	fun add(event: E, delay: Float, duration: nullable Float): EventInfo[E]
+	do
+		return add_at(event, time + delay, duration)
+	end
+
+	# Register an `event` in the queue executed at a specific time.
+	fun add_at(event: E, start: Float, duration: nullable Float): EventInfo[E]
+	do
+		if start < next then next = start
+		var ei = new EventInfo[E](event, self, start, duration or else 0.0)
+		queue.add ei
+
+		return ei
+	end
+
+	# Add a given amount of time and returns all the event that where active during the delay.
+	#
+	# Note that events that expired during the delay are marked is_dead and still returned.
+	fun update(dt: Float): SequenceRead[EventInfo[E]]
+	do
+		var time = self.time
+		time += dt
+		self.time = time
+
+		# Switch things
+		var tmp = old_actives
+		old_actives = actives
+		actives = tmp
+
+		# Discard dead events
+		actives.clear
+		for ei in old_actives do
+			if not ei.is_dead then actives.add ei
+		end
+
+		# Start new events
+		if time >= next then loop
+			if queue.is_empty then
+				next = inf
+				break
+			end
+			var ei = queue.peek
+			if ei.start > time then
+				next = ei.start
+				break
+			end
+			ei = queue.take
+			if not ei.is_dead then actives.add ei
+		end
+
+		if actives.is_empty then return actives
+
+		# Update event information
+		for ei in actives do ei.update(time, dt)
+
+		return actives
+	end
+end
+
+# Information and management about a registered event
+#
+# It allows to retrieve static and dynamic information about an event.
+# It also allows to register other time-related events with a fluent programming way.
+class EventInfo[E]
+	# The registered event
+	var event: E
+
+	# The associated event queue
+	#
+	# It is used internally for fluent programming.
+	var queue: EventQueue[E]
+
+	# Absolute start time for the registered event in the event queue time frame
+	var start: Float
+
+	# Registered duration
+	var duration: Float
+
+	# Time since the begin of the event, in units of time
+	var time: Float = nan
+
+	# Time since the last update (or the begin of the event if smaller)
+	var dt: Float = nan
+
+	# Ratio of completion
+	#
+	# Usually between 0.0 and 1.
+	# It might be > 1.0 if the event expired during the time-frame.
+	var completion = 0.0
+
+	# Is the event dead?
+	#
+	# Such an event will be not present in the next `update`.
+	var is_dead = false
```

Comment:
I don't like this method name... why not `is_expired`? It is more precise and it's the expression used in all the related documentation.